zoul: Add support for flashing to multiple devices

arthurfabre opened the pull request:
This adds support for flashing to multiple devices, by implementing the
$MOTES option used by the Z1.
This replaces the $PORT interface, as it only allowed flashing to one
device at a time, and the $MOTES interface was still used for other
tasks (such as getting a list of connected devices).

A foreach loop is used to dynamically create an upload rule per device
that needs to be flashed.
The "main" upload rule depends on the device specific ones.
This allows for parallel flashing.

arthurfabre commented:
Just found a bug with the `BSL_ADDRESS` generation, I'll push a fix when I've got it sorted.

arthurfabre commented:
Should be fixed now - there was some weird behavior with the `eval` used to create the `BSL_ADDRESS_ARG`, I've turned it into a shell variable now.

alignan commented on platform/zoul/Makefile.zoul:
```diff
@@ -69,33 +65,40 @@ else
+define UPLOAD_RULE
+%.$(MOTE): %.bin %.elf
+	@echo "Flashing $(MOTE)"
+	BSL_ADDRESS=`$(OBJDUMP) -h $$*.elf | grep -B1 LOAD | \
```
Add an `@` to avoid printing the command

alignan commented on platform/zoul/Makefile.zoul:
```diff
@@ -69,33 +65,40 @@ else
+	@# Dummy recipe to prevent "No rule to make *.upload errors"
 endif
 
+### Variable that exands into a pattern rule to upload to a given MOTE.
```
`expands`

alignan commented on platform/zoul/Makefile.zoul:
```diff
@@ -69,33 +65,40 @@ else
             perl -ne 'print $$1 . " " if(m-(/dev/\w+)-);')
 endif
 
```
````
ifdef PORT		
  MOTES = $(PORT)
endif
````

alignan commented:
I've tested and works great! :smile:

Works better if updating the `cc2538-bsl`, as the extra sorting operations adds a small delay and sometimes it overran the on-board PIC timing to put the CC2538 ready to be flashed.

I have mixed feelings as I prefer to keep the `PORT` argument, so when specified it just flashes one mote, and when `PORT` is not defined then use `MOTES`.  I know defining `MOTES=/dev/ttyUSBx` when compiling does the same effect, but this is a concern mostly based on not breaking the existing documentation and scripts relying on `PORT`, also commonly used for the cc13xx/cc26xx platforms.

Please add:
````
### If PORT is defined, override to keep backward compatibility
ifdef PORT		
  MOTES = $(PORT)
endif
````

Before this line:
````
### Check the BSL script exists
````

Thanks,, great job!

arthurfabre commented:
Updated with all the changes you mentioned :)

Also changed the `MOTES`, `MOTELIST`, and `SERIALDUMP` variables to not be recursively expanded (so make only runs the `$(shell)` command once).

alignan commented:
Sweet, thanks!

alignan commented:
:+1:
